### ISSUE TITLE
Important performance difference between training and evaluation mode

### ISSUE THREAD

**Issue by guillaumelorre28** [tensorflow/tensorflow#45911]
<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.4.0
- Python version: 3.7

Training a model using .fit method, there is an important difference of performance between training and validation even if the train split is also used for validation. The difference is the training and testing mode for the batchnorm but this should not give a difference in performance.

```python
import tensorflow as tf
import tensorflow_datasets as tfds


class Residual3x3Unit(tf.keras.layers.Layer):
    def __init__(self, channels_in, channels_out, stride, droprate=0., activate_before_residual=False):
        super(Residual3x3Unit, self).__init__()
        self.bn_0 = tf.keras.layers.BatchNormalization(momentum=0.999)
        self.relu_0 = tf.keras.layers.LeakyReLU(alpha=0.1)
        self.conv_0 = tf.keras.layers.Conv2D(channels_out, kernel_size=3, strides=stride, padding='same', use_bias=False)
        self.bn_1 = tf.keras.layers.BatchNormalization(momentum=0.999)
        self.relu_1 = tf.keras.layers.LeakyReLU(alpha=0.1)
        self.conv_1 = tf.keras.layers.Conv2D(channels_out, kernel_size=3, strides=1, padding='same', use_bias=False)
        self.downsample = channels_in != channels_out
        self.shortcut = tf.keras.layers.Conv2D(channels_out, kernel_size=1, strides=stride, use_bias=False)
        self.activate_before_residual = activate_before_residual
        self.dropout = tf.keras.layers.Dropout(rate=droprate)
        self.droprate = droprate

    def call(self, x, training=True):
        if self.downsample and self.activate_before_residual:
            x = self.relu_0(self.bn_0(x, training=training))
        elif not self.downsample:
            out = self.relu_0(self.bn_0(x, training=training))
        out = self.relu_1(self.bn_1(self.conv_0(x if self.downsample else out), training=training))
        if self.droprate > 0.:
            out = self.dropout(out)
        out = self.conv_1(out)
        return out + (self.shortcut(x) if self.downsample else x)


class ResidualBlock(tf.keras.layers.Layer):
    def __init__(self, n_units, channels_in, channels_out, unit, stride, droprate=0., activate_before_residual=False):
        super(ResidualBlock, self).__init__()
        self.units = self._build_unit(n_units, unit, channels_in, channels_out, stride, droprate, activate_before_residual)

    def _build_unit(self, n_units, unit, channels_in, channels_out, stride, droprate, activate_before_residual):
        units = []
        for i in range(n_units):
            units.append(unit(channels_in if i == 0 else channels_out, channels_out, stride if i == 0 else 1, droprate, activate_before_residual))
        return units

    def call(self, x, training=True):
        for unit in self.units:
            x = unit(x, training=training)
        return x


class WideResNet(tf.keras.Model):
    def __init__(self, num_classes, depth=28, width=2, droprate=0., input_shape=(None, 32, 32, 3), **kwargs):
        super(WideResNet, self).__init__(input_shape, **kwargs)
        assert (depth - 4) % 6 == 0
        N = int((depth - 4) / 6)
        channels = [16, 16 * width, 32 * width, 64 * width]

        self.conv_0 = tf.keras.layers.Conv2D(channels[0], kernel_size=3, strides=1, padding='same', use_bias=False)
        self.block_0 = ResidualBlock(N, channels[0], channels[1], Residual3x3Unit, 1, droprate, True)
        self.block_1 = ResidualBlock(N, channels[1], channels[2], Residual3x3Unit, 2, droprate)
        self.block_2 = ResidualBlock(N, channels[2], channels[3], Residual3x3Unit, 2, droprate)
        self.bn_0 = tf.keras.layers.BatchNormalization(momentum=0.999)
        self.relu_0 = tf.keras.layers.LeakyReLU(alpha=0.1)
        self.avg_pool = tf.keras.layers.AveragePooling2D((8, 8), (1, 1))
        self.flatten = tf.keras.layers.Flatten()
        self.dense = tf.keras.layers.Dense(num_classes, activation='softmax')

    def call(self, inputs, training=True):
        x = inputs
        x = self.conv_0(x)
        x = self.block_0(x, training=training)
        x = self.block_1(x, training=training)
        x = self.block_2(x, training=training)
        x = self.relu_0(self.bn_0(x, training=training))
        x = self.avg_pool(x)
        x = self.flatten(x)
        x = self.dense(x)
        return x

train_ds = tfds.load('cifar10', split="train")

test_ds = tfds.load('cifar10', split="train")

def preprocess(ex):

    image = tf.cast(ex['image'], tf.float32) / 255.
    return image, ex['label']

train_ds = train_ds.map(preprocess).shuffle(1024).repeat().batch(64)
test_ds = test_ds.map(preprocess).repeat().batch(64)

model = WideResNet(10)

optimizer = tf.keras.optimizers.SGD(0.03, momentum=0.9)
model.compile(optimizer=optimizer, loss=tf.keras.losses.SparseCategoricalCrossentropy(), metrics=['accuracy'])

model.fit(train_ds, epochs=10, steps_per_epoch=100, validation_data=test_ds, validation_steps=100)
```

The output is the following:

```
Epoch 1/10
100/100 [==============================] - 15s 58ms/step - loss: 1.9948 - accuracy: 0.2491 - val_loss: 3.6047 - val_accuracy: 0.1061
Epoch 2/10
100/100 [==============================] - 5s 51ms/step - loss: 1.6504 - accuracy: 0.3899 - val_loss: 2.1616 - val_accuracy: 0.2086
Epoch 3/10
100/100 [==============================] - 5s 51ms/step - loss: 1.4418 - accuracy: 0.4759 - val_loss: 5.3945 - val_accuracy: 0.1839
Epoch 4/10
100/100 [==============================] - 5s 51ms/step - loss: 1.3975 - accuracy: 0.4871 - val_loss: 2.5811 - val_accuracy: 0.2809
Epoch 5/10
100/100 [==============================] - 5s 52ms/step - loss: 1.2669 - accuracy: 0.5371 - val_loss: 2.1341 - val_accuracy: 0.3431
Epoch 6/10
100/100 [==============================] - 5s 51ms/step - loss: 1.1513 - accuracy: 0.5834 - val_loss: 3.0617 - val_accuracy: 0.3303
Epoch 7/10
100/100 [==============================] - 5s 51ms/step - loss: 1.1181 - accuracy: 0.5982 - val_loss: 2.0735 - val_accuracy: 0.4008
Epoch 8/10
100/100 [==============================] - 5s 51ms/step - loss: 1.0284 - accuracy: 0.6285 - val_loss: 3.5450 - val_accuracy: 0.3250
Epoch 9/10
100/100 [==============================] - 5s 49ms/step - loss: 1.0219 - accuracy: 0.6352 - val_loss: 1.9222 - val_accuracy: 0.4300
Epoch 10/10
100/100 [==============================] - 5s 49ms/step - loss: 0.9897 - accuracy: 0.6519 - val_loss: 2.6546 - val_accuracy: 0.4075
```

**Comment by amahendrakar** [comment]
@guillaumelorre28,
I did not face any issues while running the code with TF v2.4. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/22ec2955a2808e556b211932e6baeff5/45911.ipynb). 

Could you please let us know the expected behavior of the code, so that we can look into it? Thanks!
